src/hotspot/cpu/x86/stubGenerator_x86_64.cpp

```diff
@@ -5385,6 +5385,471 @@ address generate_avx_ghash_processBlocks() {
+    __ vpshufd(zmm_bVec, zmm_bVec, 0x39, Assembler::AVX_512bit);
+    __ vpshufd(zmm_cVec, zmm_cVec, 0x4E, Assembler::AVX_512bit);
+    __ vpshufd(zmm_dVec, zmm_dVec, 0x93, Assembler::AVX_512bit);
+
+    // The second set of operations on the vectors covers the second 4 quarter
+    // round operations, now acting on the diagonals:
+    //  Qround(state, 0, 5,10,15)
+    //  Qround(state, 1, 6,11,12)
+    //  Qround(state, 2, 7, 8,13)
+    //  Qround(state, 3, 4, 9,14)
+
+    // a += b; d ^= a; d <<<= 16
+    __ vpaddd(zmm_aVec, zmm_aVec, zmm_bVec, Assembler::AVX_512bit);
+    __ vpxor(zmm_dVec, zmm_dVec, zmm_aVec, Assembler::AVX_512bit);
+    __ evprold(zmm_dVec, zmm_dVec, 16, Assembler::AVX_512bit);
+
+    // c += d; b ^= c; b <<<= 12 (b << 12 | scratch >>> 20)
+    __ vpaddd(zmm_cVec, zmm_cVec, zmm_dVec, Assembler::AVX_512bit);
+    __ vpxor(zmm_bVec, zmm_bVec, zmm_cVec, Assembler::AVX_512bit);
+    __ evprold(zmm_bVec, zmm_bVec, 12, Assembler::AVX_512bit);
+
+    // a += b; d ^= a; d <<<= 8 (d << 8 | scratch >>> 24)
+    __ vpaddd(zmm_aVec, zmm_aVec, zmm_bVec, Assembler::AVX_512bit);
+    __ vpxor(zmm_dVec, zmm_dVec, zmm_aVec, Assembler::AVX_512bit);
+    __ evprold(zmm_dVec, zmm_dVec, 8, Assembler::AVX_512bit);
+
+    // c += d; b ^= c; b <<<= 7 (b << 7 | scratch >>> 25)
+    __ vpaddd(zmm_cVec, zmm_cVec, zmm_dVec, Assembler::AVX_512bit);
+    __ vpxor(zmm_bVec, zmm_bVec, zmm_cVec, Assembler::AVX_512bit);
+    __ evprold(zmm_bVec, zmm_bVec, 7, Assembler::AVX_512bit);
+
+    // Before we start the next iteration, we need to perform shuffles
+    // on the b/c/d vectors to move them back to columnar organizations
+    // from their current diagonal orientation.
+    __ vpshufd(zmm_bVec, zmm_bVec, 0x93, Assembler::AVX_512bit);
+    __ vpshufd(zmm_cVec, zmm_cVec, 0x4E, Assembler::AVX_512bit);
+    __ vpshufd(zmm_dVec, zmm_dVec, 0x39, Assembler::AVX_512bit);
+
+    __ decrement(loopCounter);
+    __ jcc(Assembler::notZero, L_twoRounds);
+
+    // Add the initial state to the end register values
+    // We use a scratch register to replicate the 128-bit
+    // state data for the A/B/C/D vectors.  We will also add in
+    // the counter add mask onto zmm3 after adding in the start state.
+    __ evbroadcasti32x4(zmm_scratch, Address(state, 0), Assembler::AVX_512bit);
+    __ vpaddd(zmm_aVec, zmm_aVec, zmm_scratch, Assembler::AVX_512bit);
+    __ evbroadcasti32x4(zmm_scratch, Address(state, 16), Assembler::AVX_512bit);
+    __ vpaddd(zmm_bVec, zmm_bVec, zmm_scratch, Assembler::AVX_512bit);
+    __ evbroadcasti32x4(zmm_scratch, Address(state, 32), Assembler::AVX_512bit);
+    __ vpaddd(zmm_cVec, zmm_cVec, zmm_scratch, Assembler::AVX_512bit);
+    __ evbroadcasti32x4(zmm_scratch, Address(state, 48), Assembler::AVX_512bit);
+    __ vpaddd(zmm_dVec, zmm_dVec, zmm_scratch, Assembler::AVX_512bit);
+    __ vpaddd(zmm_dVec, zmm_dVec, ExternalAddress(StubRoutines::x86::chacha20_counter_addmask_avx512()), Assembler::AVX_512bit, rax);
```

Comment:
These could directly use the values in zmm_s1 to zmm_s5 registers  :
    __ vpaddd(zmm_aVec, zmm_aVec, zmm_s1, Assembler::AVX_512bit);
    ...
    __ vpaddd(zmm_dVec, zmm_dVec, zmm_s5, Assembler::AVX_512bit);

Reply:
Keeping the original broadcasted state data on registers was a good idea, as it saved me the extra reach out to memory at the end of the loop.  Fixed as recommended.